chain/epoch-manager/src/shard_tracker.rs

```diff
@@ -340,6 +346,131 @@ impl ShardTracker {
             Ok(Some(state_sync_info))
         }
     }
+
+    /// Checks whether the given `shard_id` is part of the scheduled tracked shards for the
+    /// specified `epoch_id`, based on the provided `schedule`.
+    fn check_if_shard_is_tracked_according_to_schedule(
+        &self,
+        shard_id: ShardId,
+        schedule: &Vec<Vec<ShardId>>,
+        epoch_id: &EpochId,
+    ) -> Result<bool, EpochError> {
+        assert_ne!(schedule.len(), 0);
+        let epoch_info = self.epoch_manager.get_epoch_info(epoch_id)?;
+        let epoch_height = epoch_info.epoch_height();
+        let index = epoch_height % schedule.len() as u64;
+        let subset = &schedule[index as usize];
+        Ok(subset.contains(&shard_id))
+    }
+
+    /// Checks whether `shard_id` contains any of the `tracked_accounts` in the given `epoch_id`.
+    fn check_if_shard_contains_tracked_account(
+        &self,
+        shard_id: ShardId,
+        tracked_accounts: &Vec<AccountId>,
+        epoch_id: &EpochId,
+    ) -> Result<bool, EpochError> {
+        let shard_layout = self.epoch_manager.get_shard_layout(epoch_id)?;
+        let tracking_mask = self.tracked_accounts_shard_cache.get_or_try_put(
+            *epoch_id,
+            |_| -> Result<Vec<bool>, EpochError> {
+                let mut tracking_mask = shard_layout.shard_ids().map(|_| false).collect_vec();
+                for account_id in tracked_accounts {
+                    let shard_id = shard_layout.account_id_to_shard_id(account_id);
+                    let shard_index = shard_layout.get_shard_index(shard_id)?;
+                    tracking_mask[shard_index] = true;
+                }
+                Ok(tracking_mask)
+            },
+        )?;
+        let shard_index = shard_layout.get_shard_index(shard_id)?;
+        Ok(tracking_mask.get(shard_index).copied().unwrap_or(false))
+    }
+
+    /// Checks whether `shard_id` is a descendant of any of the `tracked_shards`.
+    /// Assumes that `shard_id` exists in the shard layout of `epoch_id`.
+    pub fn check_if_descendant_of_tracked_shard(
+        &self,
+        shard_id: ShardId,
+        tracked_shards: &Vec<ShardUId>,
+        epoch_id: &EpochId,
+    ) -> Result<bool, EpochError> {
+        if let Some(is_tracked) = self.descendant_of_tracked_shard_cache.lock().get(&shard_id) {
+            return Ok(*is_tracked);
+        }
+
+        let is_tracked = check_if_descendant_of_tracked_shard_impl(
+            shard_id,
+            &tracked_shards,
+            &epoch_id,
+            &self.epoch_manager,
+        )?;
+        self.descendant_of_tracked_shard_cache.lock().insert(shard_id, is_tracked);
+        Ok(is_tracked)
+    }
+
+    pub fn get_shards_tracks_in_epoch_non_validator(
```

Comment:
nit: Can you add a comment please? Especially the `non_validator` part isn't clear to me. 
Also grammar is a bit off but I don't mind given it clearly points to the `tracks_shard_at_epoch`